Add ThirdReality Smart Color Bulb ZL1 (3RCB01057Z)

ekobres opened the pull request:
## Proposed change
<!--
  Explain your proposed change below.
-->
Add quirk for ThirdReality Smart Color Bulb ZL1 (3RCB01057Z)

## Additional information
<!--
  Please include any additional information that is important to this PR.
  For example, if this PR is a potentially breaking change, mention that here.
  If this PR requires other PRs to be merged in HA Core or other projects, mention that.
  Lastly, if this PR fixes a specific issue, please include "Fixes #xxxx".
-->
This is a new quirk. No breaking changes.

This quirk fixes a few problems with the ZL1 Bulb:

1.  Correct logical color temperature presentation (device reports 142-454
       mireds but is actually 154-370). Map between logical and device ranges.
2.  Provide perceptual brightness mapping so very low but non-off levels
       (raw 1..) are reachable and high end is slightly compressed.
3.  Snap XY and Hue/Sat commands to the nearest CT in u'v' when close,
       emitting MoveToColorTemp with default transition; expose logical CT
       154–370 while mapping to device 142–454. (So "Hey Siri, set the
       lights to X" will work if you previously asked what color a light was
       while it was in CT mode and want to set it back to that temperature.)

## Device diagnostics
<!--
  Diagnostics data is used by ZHA unit tests to make sure quirks create the expected
  entities and we do not have regressions. For all new quirks, we require device
  diagnostics.

  You can find the diagnostics information by going to the device page, clicking the
  three dots, and then by clicking on "Download diagnostics". Drag-and-drop the
  downloaded file into this section.
-->
[zha-1c4d17a122adeb322c60c77f5939f0ee-Third Reality, Inc 3RCB01057Z-563d149f0a3ab26a9bc0e2d9f377fe39.json](https://github.com/user-attachments/files/22244358/zha-1c4d17a122adeb322c60c77f5939f0ee-Third.Reality.Inc.3RCB01057Z-563d149f0a3ab26a9bc0e2d9f377fe39.json)

## Checklist
<!--
  Put an 'x' in all boxes that apply.
  Note: You do not need to tick all boxes before creating a PR.
-->

- [X] The changes are tested and work correctly
- [X] `pre-commit` checks pass / the code has been formatted using Black
- [X] Tests have been added to verify that the new code works
- [X] Device diagnostics data has been attached

ekobres committed:
thirdreality: ZL1 quirk: fix CT mapping (142–454→154–370), u’v’ CT snapping for XY/HS, and perceptual brightness with anti-OFF

puddly commented:
Interesting PR! Superifically, these look like firmware problems.

@3reality-support are these issues you can address?

ekobres committed:
tests: add Third Reality ZL1 quirk test suite

Adds tests/test_thirdreality_zl1.py covering the ZL1 color/level quirk.\n- CT normalization (device 142–454 ↔ logical 154–370) and bounds injection.\n- u’v’-based snapping for MoveToColor/MoveToHueSat to CT; XY passthrough off-locus.\n- color_mode updates, parse fallbacks, and unhandled-command passthroughs.\n- LevelControl perceptual mapping, anti-off, hysteresis, and LUT fallbacks.\n- Bind seeding and CT u’v’ cache reuse.\n- ~85% coverage for the quirk module.

ekobres commented:
> Interesting PR! Superifically, these look like firmware problems.
> 
> @3reality-support are these issues you can address?

They are firmware issues to be sure. Firmware fixes would be preferable, but I haven't gotten any traction on their support forums. In the meantime, this quirk has my lights behaving properly. If they are listening and respond, I have suggestions...

To add - if they do patch these changes in firmware it's likely to pretty badly break scenes people have already created since their brightness and CT levels will be noticeably (albeit correctly) changed from whatever they have been living with.

ekobres commented:
Hi maintainers,

The failing test (test_xiaomi_eu_plug_binding[PlugMAEU01] in tests/test_xiaomi.py) is not related to the changes in this PR. It appears to be failing in CI for unrelated reasons (see job: https://github.com/zigpy/zha-device-handlers/actions/runs/17599529930/job/49998865778?pr=4335).

Could you please:

Confirm this test failure is unrelated to my changes, and/or
Rerun the workflow, or
Allow merging despite this unrelated failure?
Thank you!

ekobres committed:
ci: retrigger 3.13 job after confirming upstream dev passes

ekobres committed:
Merge branch 'dev' into feature/thirdreality-zl1-quirk

3reality-support commented:
> > Interesting PR! Superifically, these look like firmware problems.
> > @3reality-support are these issues you can address?
> 
> They are firmware issues to be sure. Firmware fixes would be preferable, but I haven't gotten any traction on their support forums. In the meantime, this quirk has my lights behaving properly. If they are listening and respond, I have suggestions...
> 
> To add - if they do patch these changes in firmware it's likely to pretty badly break scenes people have already created since their brightness and CT levels will be noticeably (albeit correctly) changed from whatever they have been living with.

@ekobres 1. Regarding (device reports 142-454mireds  but actually 154-370). The report is indeed 142-454, but how did you obtain 154-370? Is it the device interface?
2. Is it not dark enough when the brightness is 1%?
3. At present, we have achieved that when setting xy or hs, if it approaches the range of color temperature, it will transition to color temperature

ekobres commented:
> @ekobres 1. Regarding (device reports 142-454mireds but actually 154-370). The report is indeed 142-454, but how did you obtain 154-370? Is it the device interface? 

@3reality-support Thank you for responding!

1. I got 154-370 (2700K-6500K) from your [web site](https://3reality.com/product/smart-color-bulb-zl1), your [user manual](https://3reality.com/wp-content/uploads/2024/10/Smart-Color-Bulb-ZL1_UM_20240924.pdf), 

your product packaging:...

![](https://3reality.com/wp-content/uploads/2024/09/Smart-Color-Bulb-ZL1.jpg) 

and the bulb itself:

![IMG_3030](https://github.com/user-attachments/assets/4a56a71a-2825-4c48-91dd-8c03c8697f0d)

Also, empirical testing next to 4 other smart and dumb bulbs that match at 2700K. When I set your CT to 454 mired, it is a perfect match for 2700K. When I set it at 143 mired, it matches 6500K. So it was a pretty safe bet I could map your 311 mired range into the normal 216 mired range of 2700K-6500K. That part works beautifully and with my quirk, your bulbs behave just exactly like my other tunable CT bulbs from 154-370.

> 2. Is it not dark enough when the brightness is 1%? 

I did some informal testing with a lux meter held to the diffuser of your bulb, and a Hue ambience bulb. I know you don't have the hardware to match Hue's very dim settings, and that is okay. But it also does not appear that you build a perceptual curve into your brightness. Human brightness perception is logarithmic - everyone can immediately see the difference between 10-20 lux. It's much harder to tell the difference between 1,000 and 2,000 lux. It's even harder still to tell 10,000 from 20,000 lux - and so on. So I did my best to squash your low-end brightness and stay idempotent on the round trip between brightness percentage and raw brightness settings (1-255). There is only so much correction I can do in a quirk since I am locked into your 255 possible values - but by flattening the bottom 50% of the HomeAssistant 1-255 range and then compressing on a curve to meet at 254 - it's a crude approximation of a perceptual curve, but it's far better than the default in the firmware.

I'm sure you could do much, much better in firmware by programming a true brightness perceptual curve into a 255-element level table. 

> 3. At present, we have achieved that when setting xy or hs, if it approaches the range of color temperature, it will transition to color temperature.

I would love to test this. I borrowed the math from the internet, so I can't take credit, but I am very happy with my quirk's .015 epsilon u'v' distance snap to the blackbody curve. This matches several Apple colors exactly the same as Hue - including a perfect 370 mired "Topaz" color and a nice 360 mired "Desert Sand" color - and "Ivory", and a few others. It would be very handy to have a manufacturer cluster attribute to set the epsilon so that the snap could be tuned.

Another request: Please add your color gamut RGB x/y/intensity and primary driver 1-5 x/y/intensity attributes in 0x0300: 0x11 - 0x3B. This can help a lot with calibrating colors with real objects and with other types of lights. Also, it would be nice if your bulb would clamp XY at the actual CIE 1931 XY area of your bulb. It would also be amazing if you could publicly share the true color gamut chart and the photo spectrograph information for your bulbs. Even if you decide not to clamp or populate the RGB X/Y/I and primary XY/I, having the real gamut would allow us to achieve these things with quirks.

Last thing - is it possible to share what was changed in the 1.00.74 firmware that was released recently?

I really do hope you can fix all of this in firmware!

Please let me know if you have any questions, or let me know how I can help!

Off topic: 

I really like your products, and I know the community is excited for you to continue developing new lights and other products. 
I pre-ordered your dual-receptacle energy-monitoring plug, and I am looking forward to the things I can do with it!

ekobres committed:
thirdreality: trim unused paths, strengthen tests to cover error/edge branches; pre-commit clean (ruff/mypy/format)

ekobres commented:
Had to rebase - looks like the Tuya tests are no longer blocking.  (Thanks!)
In the meantime I pruned some unnecessary code and added additional tests to hit the coverage target. 

This PR is ready for review.

ekobres commented on zhaquirks/thirdreality/smart_color_bulb_zl1.py:
```diff
@@ -0,0 +1,1258 @@
+            return 0.3127, 0.3290
+        return x_val / denom, y_val / denom
+
+    @staticmethod
+    def _inv_gamma(u: float) -> float:
+        return (u / 12.92) if u <= 0.04045 else (((u + 0.055) / 1.055) ** 2.4)
```
For brevity. Interested parties can look them up.

ekobres commented on zhaquirks/thirdreality/smart_color_bulb_zl1.py:
```diff
@@ -0,0 +1,1258 @@
+        T = float(cct)
+        T = max(T, 1667)
+        T = min(T, 25000)
+        if T <= 4000:
+            x = (
+                -0.2661239 * (1e9 / (T**3))
+                - 0.2343589 * (1e6 / (T**2))
+                + 0.8776956 * (1e3 / T)
+                + 0.179910
+            )
+        else:
+            x = (
+                -3.0258469 * (1e9 / (T**3))
+                + 2.1070379 * (1e6 / (T**2))
+                + 0.2226347 * (1e3 / T)
+                + 0.240390
+            )
```
For brevity. Interested parties can look them up.

ekobres commented on zhaquirks/thirdreality/smart_color_bulb_zl1.py:
```diff
@@ -0,0 +1,1258 @@
+        val = min(val, 1.0)
+        return int(round(val * 65535.0))
+
+    @staticmethod
+    def _xy_from_cct(cct: float):
+        T = float(cct)
+        T = max(T, 1667)
+        T = min(T, 25000)
+        if T <= 4000:
+            x = (
+                -0.2661239 * (1e9 / (T**3))
+                - 0.2343589 * (1e6 / (T**2))
+                + 0.8776956 * (1e3 / T)
+                + 0.179910
+            )
+        else:
+            x = (
+                -3.0258469 * (1e9 / (T**3))
+                + 2.1070379 * (1e6 / (T**2))
+                + 0.2226347 * (1e3 / T)
+                + 0.240390
+            )
+        if T <= 2222:
+            y = -1.1063814 * (x**3) - 1.34811020 * (x**2) + 2.18555832 * x - 0.20219683
+        elif T <= 4000:
+            y = -0.9549476 * (x**3) - 1.37418593 * (x**2) + 2.09137015 * x - 0.16748867
+        else:
+            y = 3.0817580 * (x**3) - 5.87338670 * (x**2) + 3.75112997 * x - 0.37001483
```
For brevity. Interested parties can look them up.

ekobres commented on zhaquirks/thirdreality/smart_color_bulb_zl1.py:
```diff
@@ -0,0 +1,1258 @@
+
+    @staticmethod
+    def _cct_from_xy(x: float, y: float) -> float:
+        n = (x - 0.3320) / (y - 0.1858) if (y - 0.1858) != 0 else 0.0
+        cct = 449 * (n**3) + 3525 * (n**2) + 6823.3 * n + 5520.33
```
For brevity. Interested parties can look them up.

ekobres commented on zhaquirks/thirdreality/smart_color_bulb_zl1.py:
```diff
@@ -0,0 +1,1258 @@
+            r1, g1, b1 = C, 0.0, x_comp
+        # Inverse gamma to linear (use static helper to avoid scoping confusion)
+        rl, gl, bl = self._inv_gamma(r1), self._inv_gamma(g1), self._inv_gamma(b1)
+        x_val = 0.4124 * rl + 0.3576 * gl + 0.1805 * bl
+        y_val = 0.2126 * rl + 0.7152 * gl + 0.0722 * bl
+        z_val = 0.0193 * rl + 0.1192 * gl + 0.9505 * bl
```
For brevity. Interested parties can look them up.
